### PR TITLE
Migrate the API to MuData

### PR DESCRIPTION
**Pull request by alihamraoui**
This PR migrates CustOmics from the `(dict[str, DataFrame], clinical_df)` input convention to a single [MuData](https://mudata.readthedocs.io/) object, the standard container for multi-omics data.
One object now carries every omics modality (as `AnnData` modalities in `mdata.mod`) and the clinical annotations (in `mdata.obs`), so the per-call `clinical_df` / `label` / `event` / `surv_time` arguments are gone.

## What changed

- **`fit`, `evaluate`, `predict`, `predict_survival`, `explain`, `plot_representation`** now take a
  single `mdata: MuData` argument instead of `(omics_df, clinical_df, label,
  event, surv_time)`.
- **New `customics.prepare_input(mdata, label, event, surv_time)`** validates the
  clinical columns exist in `mdata.obs` and registers them in `mdata.uns` (under
  `customics_label`, `customics_event`, `customics_surv_time`). `fit`/`evaluate`
  read the targets from there, so you specify them once.
- **`get_common_samples`** now aligns on `AnnData.obs_names` instead of DataFrame
  indices.
- **`toy_dataset()`** now returns a ready-to-use `MuData` instead of a
  `(dict, DataFrame)` tuple.
- `mudata` / `anndata` added as dependencies; tests and fixtures updated.

## New usage

```python
import torch
import customics
from customics import CustOMICS, prepare_input

# 1. Load data as a MuData object (modalities in .mod, clinical in .obs)
mdata = customics.toy_dataset()

# 2. Register the clinical targets once
prepare_input(mdata, label="cluster.id", event="OS", surv_time="OS.time")

# 3. Configure & train
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model = CustOMICS(
    source_params=source_params,
    central_params=central_params,
    classif_params=classif_params,
    surv_params=surv_params,
    train_params=train_params,
    device=device,
)
model.fit(mdata, n_epochs=30, batch_size=32)

# 4. Evaluate (targets are read from mdata.uns)
acc = model.evaluate(mdata, task="classification")
ci  = model.evaluate(mdata, task="survival")
```

### Todo list:

- [ ]  README and tutorial docs to be updated to the MuData API.

### CODE REVIEW

File: customics/datasets/multi_omics_dataset.py [hunk 1]
```diff
@@ -14,8 +15,8 @@ class MultiOmicsDataset(Dataset):
 
     Parameters
     ----------
-    omics_df : dict of str → pd.DataFrame
-        Multi-omics data.  Each DataFrame must be indexed by sample ID.
+    mdata : MuData
+            Multi-omics object.
     clinical_df : pd.DataFrame
```
Comment (quentinblampey) [inline customics/datasets/multi_omics_dataset.py]: We don't need the `clinical_df` anymore, no? Since you have it in the `.obs`?

I think you can move the logic of the `encoded_clinical` creation within the `MultiOmicsDataset` (but let me know if I'm wrong)

File: customics/model.py [hunk 2]
```diff
@@ -763,9 +747,9 @@ def explain(
         )
 
         self._require_fitted()
-        expr_df = omics_df[source]
+        expr_df = mdata[source].to_df()
         sample_id = list(set(sample_id) & set(expr_df.index))
-        phenotype = processPhenotypeDataForSamples(clinical_df, sample_id, self.label_encoder)
+        phenotype = processPhenotypeDataForSamples(mdata.obs, sample_id, self.label_encoder)
```
Comment (quentinblampey) [inline customics/model.py]: Not related to your PR, but I see that `processPhenotypeDataForSamples` is not written in snake_case

This is the same for other functions in the `explain/shap.py` file

File: customics/utils.py [hunk 3]
```diff
@@ -39,30 +43,64 @@ def toy_dataset() -> tuple[dict[str, pd.DataFrame], pd.DataFrame]:
...
 
 
 # ---------------------------------------------------------------------------
 # Sample alignment
 # ---------------------------------------------------------------------------
 
 
-def get_common_samples(dfs: list[pd.DataFrame]) -> list[str]:
-    """Return sample IDs present in every DataFrame.
+def get_common_samples(adatas: list[AnnData]) -> list[str]:
```
Comment (quentinblampey) [inline customics/utils.py]: I think you can provide directly a `MuData` object as an input for conveniency

File: customics/visualization.py [hunk 4]
```diff
@@ -128,22 +123,22 @@ def plot_survival_stratification(
...
     median_hazard = np.mean(hazard_pred)
     high = [s for s, h in zip(lt_samples, hazard_pred) if h > median_hazard]
     low = [s for s, h in zip(lt_samples, hazard_pred) if h <= median_hazard]
 
     kmf_low = KaplanMeierFitter(label="low risk")
     kmf_high = KaplanMeierFitter(label="high risk")
-    kmf_low.fit(clinical_df.loc[low, surv_time], clinical_df.loc[low, event])
-    kmf_high.fit(clinical_df.loc[high, surv_time], clinical_df.loc[high, event])
+    kmf_low.fit(mdata.obs.loc[low, surv_time], mdata.obs.loc[low, event])
+    kmf_high.fit(mdata.obs.obsal_df.loc[high, surv_time], mdata.obs.loc[high, event])
```
Comment (quentinblampey) [inline customics/visualization.py]: I didn't know `mdata.obs.obsal_df`! What is it (out of curiosity)?

File: customics/model.py [hunk 5]
```diff
@@ -464,21 +462,18 @@ def fit(
 
     def _validate_fit_inputs(
         self,
-        omics_train: dict[str, pd.DataFrame],
-        clinical_df: pd.DataFrame,
-        label: str,
-        event: str,
-        surv_time: str,
+        mdata: MuData,
+        clinical_params: list,
```
Comment (quentinblampey) [inline customics/model.py]: I think you can remove the `clinical_params` argument and always use `[Keys.LABEL, Keys.EVENT, Keys.SURV_TIME]` instead.

Also, you could return `label, event, surv_time`, this way you can directly get them in .fit() without adding three lines afterward.

File: customics/datasets/multi_omics_dataset.py [hunk 6]
```diff
@@ -56,7 +57,9 @@ def __len__(self) -> int:
 
     def __getitem__(self, index: int) -> tuple[list[torch.Tensor], int, int, int]:
         sample = self.lt_samples[index]
-        omics_data = [torch.tensor(df.loc[sample, :].values.astype(np.float32)) for df in self.omics_df.values()]
+        omics_data = [
+            torch.tensor(adata.to_df().loc[sample, :].values.astype(np.float32)) for adata in self.mdata.mod.values()
```
Comment (quentinblampey) [inline customics/datasets/multi_omics_dataset.py]: Slight optimization (but not very important because these are small objects):
When you do `adata.to_df().loc[sample, :].values`, you create a dataframe with all samples and then you subset. I think it would be more efficient to first subset and then convert to a dataframe no?
E.g., something like `adata[sample].X[0]` (I didn't test it though)

File: customics/utils.py
Comment (quentinblampey) [inline customics/utils.py]: Minor: I think we don't need to make a DataFrame anymore because we de-structure it just after. You can make three variables instead, e.g. `protein_df = ...`, `gene_exp_df = ...`, `methyl_df = ...`

File: customics/utils.py [hunk 7]
```diff
@@ -39,30 +43,64 @@ def toy_dataset() -> tuple[dict[str, pd.DataFrame], pd.DataFrame]:
...
+        (1 = event, 0 = censored).
+    surv_time : str
+        Name of the ``mdata.obs`` column holding the survival time.
+
+    Raises
+    ------
+    DataValidationError
+        If any of the given columns is missing from ``mdata.obs``.
+    """
+    for key, column in {"label": label, "event": event, "surv_time": surv_time}.items():
```
Comment (quentinblampey) [inline customics/utils.py]: You can directly use `Keys.LABEL` instead of `"label"` (and the same goes for the other two): this way you avoid hardcoding the strings, and you can replace the afteward `f"customics_{key}"` by simply `key`

File: tests/integration/test_training.py [hunk 8]
```diff
@@ -13,8 +13,7 @@ def test_full_pipeline_classification(
     surv_params,
     train_params,
     device,
-    omics_df,
-    clinical_df,
+    mu_data,
```
Comment (quentinblampey) [inline tests/integration/test_training.py]: Usually we call it `mdata` and not `mu_data`

**Review by quentinblampey — CHANGES_REQUESTED** [review]
Thanks @alihamraoui, this is really awesome!
I made some comments, but everything is very minor (mostly code quality and readability)

File: customics/datasets/multi_omics_dataset.py (hunk 1 above)
Comment (alihamraoui) [reply customics/datasets/multi_omics_dataset.py]: Great point! I actually tried that, but `LabelEncoder` is fit inside `fit()` and then reused by the validation loader and by `evaluate()`.

What do you think about passing an optional, already-fitted `label_encoder` to the `MultiOmicsDataset`?

OR, simpler: keep the encoder on the model and just pass the encoded label Series to `MultiOmicsDataset`, this also saves memory by avoiding the `df.copy()`.

File: customics/datasets/multi_omics_dataset.py (hunk 6 above)
Comment (alihamraoui) [reply customics/datasets/multi_omics_dataset.py]: Awesome! that makes sense.

File: customics/datasets/multi_omics_dataset.py (hunk 1 above)
Comment (quentinblampey) [reply customics/datasets/multi_omics_dataset.py]: Yes you can try this option! You can also store the labeled Series in the obs with a constant column name, and recover it from the `MultiOmicsDataset`
